src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -193,34 +198,62 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
             Checksum solutionChecksum,
             bool fromPrimaryBranch,
             int workspaceVersion,
+            ProjectId? projectId,
             CancellationToken cancellationToken)
         {
-            var availableSolution = TryGetAvailableSolution(solutionChecksum);
-            if (availableSolution != null)
+            if (projectId == null)
             {
-                return availableSolution;
-            }
+                var availableSolution = TryGetAvailableSolution(solutionChecksum);
+                if (availableSolution != null)
+                    return availableSolution;
 
-            // make sure there is always only one that creates a new solution
-            using (await _availableSolutionsGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                // make sure there is always only one that creates a new solution
+                using (await _availableSolutionsGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    availableSolution = TryGetAvailableSolution(solutionChecksum);
+                    if (availableSolution != null)
+                        return availableSolution;
+
+                    var solution = await CreateSolution_NoLockAsync(
+                        assetProvider,
+                        solutionChecksum,
+                        shouldCacheSolution: fromPrimaryBranch,
+                        workspaceVersion,
+                        CurrentSolution,
+                        cancellationToken).ConfigureAwait(false);
+
+                    _lastRequestedSolutionWithChecksum = Tuple.Create(solutionChecksum, solution);
+                    return solution;
+                }
+            }
+            else
             {
-                availableSolution = TryGetAvailableSolution(solutionChecksum);
-                if (availableSolution != null)
+                if (_lastRequestedProjectIdToSolutionWithChecksum.TryGetValue(projectId, out (Checksum checksum, Solution solution) value) &&
+                    value.checksum == solutionChecksum)
                 {
-                    return availableSolution;
+                    return value.solution;
                 }
 
-                var solution = await CreateSolution_NoLockAsync(
-                    assetProvider,
-                    solutionChecksum,
-                    fromPrimaryBranch,
-                    workspaceVersion,
-                    CurrentSolution,
-                    cancellationToken).ConfigureAwait(false);
+                using (await _availableProjectsGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
>_availableProjectsGate [](http://example.com/codeflow?start=29&length=22)

This is odd - CreateSolution_NoLockAsync here is called under a different lock (_availableProjectsGate) then above (_availableSolutionsGate). Why do we need two locks?

Reply:
The combination of concurrent dictionary and double-checked lock seems unnecessary. Why not just Dictionary and a simple lock? The solution should only be acquired once per service request. That doesn't seem very contentious.

Reply:
i'm ok with taht.  this was to follow the existing pattern though of "checking if hte value is in the cache, if not, take lock, check again, if not htere compute and store value".  to do the check outside the lock we need a threadsafe data store.  For the non-project state we use a volatile-tuple.  For the project state i implemented that witha  concurrent dictionary.  however, as above, i'm fine changing this.

Reply:
> Why do we need two locks?

This is a good question, bit there's an important reason for it. 

We do not want the lock taken for syncing the whole solution to block is from syncing a particular project cone. 

However I agree this is confusing and hard to reason about. I'll try to refactoring this to be clearer abs to share where reasonable. 

In particular, I am likely to break out the parts of the work that produce the snapshot with the logic to mutate the workspace state. 

This pr was just to get a sense of how possible this was :-)